File: backend/app/api/collab.py
# Built with Spec4 AI - https://spec4.ai
"""The multi-agent collaboration app's HTTP surface.

Two endpoints. `GET /identity-cards` serves the three A2A-shaped cards
from module constants -- no model, no database, no quota, and nothing a visitor
supplies. It is the analogue of the orchestrated app's `/roster`, and it is
served from the API for the same reason: the cards are what the run is
validated against server-side, so a frontend copy would be a second source of
truth free to drift from the one the negotiation actually uses.

Serialisation is by alias, so the response carries A2A's own camelCase spelling
(`protocolVersion`, `toolAccess`, `defaultInputModes`). That is asserted in the
endpoint's test rather than trusted to configuration -- an alias generator that
silently stopped applying would still return a perfectly valid-looking body.

`POST /run` streams the negotiation, one event per stage. It is a stream rather
than a JSON response because the two bidding stages are concurrent and take
seconds each: the point of the demonstration is that both seller columns are
visibly in progress together, and a response assembled at the end would show a
visitor the result of parallelism without the parallelism.

**This router is thin on purpose.** Every decision -- validation, the hourly
gate, the allowance hold, the six stages, the post-checks, persistence -- lives
in the slice service. What is here is the request body, the event encoding, and
the disconnect handling.

Like the planning and orchestrated run endpoints, `/run` deliberately does not
use `Depends(get_db_session)`: a dependency's session is bound to the request
scope and this response outlives the handler, because the generator runs while
the body streams. The session is opened inside the generator instead.
"""

from __future__ import annotations

import asyncio
import json
import uuid
from collections.abc import AsyncGenerator
from typing import Any

import structlog
from fastapi import APIRouter
from pydantic import BaseModel, Field

# Imported from the package root rather than `sse_starlette.sse`: only the
# root declares these in `__all__`, and this module is type-checked strictly
# (unlike the pre-v6 routers, which inherit an exemption).
from sse_starlette import EventSourceResponse, ServerSentEvent

from backend.app.collab.protocol import AgentCard
from backend.app.collab.scenarios import IDENTITY_CARDS
from backend.app.collab.service import stream_run
from backend.app.db.session import async_session_factory

logger = structlog.get_logger()

router = APIRouter(prefix="/api/collab", tags=["collab"])

#: Seconds between keep-alive comments. Render's proxy closes an idle
#: connection, and a bidding stage against a free model can run for tens of
#: seconds with nothing to say.
PING_SECONDS = 15


class IdentityCard(BaseModel):
    """One agent's published card, with what the bus and the UI need alongside it.

    `id`, `role` and `color` are this app's, not A2A's -- an `AgentCard`
    describes an agent rather than addressing one. Keeping them beside the card
    rather than inside it is what lets `protocol.py` stay a faithful statement
    of the protocol's shape.

    Attributes:
        id: The agent's address on the message bus.
        role: `buyer` or `seller`.
        color: Accent for this agent's track.
        card: What the peer publishes about itself.
    """

    id: str
    role: str
    color: str
    card: AgentCard


class IdentityCardsResponse(BaseModel):
    """Response body of `GET /api/collab/identity-cards`.

    Attributes:
        agents: The three participants, buyer first.
    """

    agents: list[IdentityCard]


@router.get("/identity-cards", response_model_by_alias=True)
async def get_identity_cards() -> IdentityCardsResponse:
    """Return the three peer identity cards for inspection.

    Static and free: this is what each party publishes about itself, readable
    before a run starts and while one is in flight. Private negotiating
    positions are not here and are not derivable from what is -- a card is the
    public face, and the sealed constraints stay sealed until the run ends.

    Returns:
        The buyer and both rival sellers, each with its A2A-shaped card.
    """
    return IdentityCardsResponse(
        agents=[
            IdentityCard(
                id=agent.id,
                role=agent.role,
                color=agent.color,
                card=agent.card,
            )
            for agent in IDENTITY_CARDS
        ]
    )


class RunRequest(BaseModel):
    """Body of `POST /api/collab/run`.

    Both fields are closed sets: a scenario id from the catalogue and either a
    preset weighting id or an explicit numeric vector. **No free text reaches
    this app at all**, which is why it needs no moderation gate -- the one
    example in the showcase that genuinely does not.

    Attributes:
        scenario_id: The scenario the visitor chose.
        weighting_id: A preset weighting id.
        weights: An explicit per-axis vector, when not using a preset.
    """

    scenario_id: str = Field(min_length=1, max_length=64)
    weighting_id: str | None = Field(default=None, max_length=64)
    weights: dict[str, int] | None = None


def _event(name: str, payload: dict[str, Any]) -> ServerSentEvent:
    """Encode one stage event as a named SSE event.

    Args:
        name: The event name the client listens for.
        payload: JSON-serialisable body.

    Returns:
        The encoded event.
    """
    return ServerSentEvent(event=name, data=json.dumps(payload))


@router.post("/run")
async def run(payload: RunRequest) -> EventSourceResponse:
    """Start a negotiation and stream one event per stage.

    A 200 with the refusal carried as an `error` event, following the
    convention the planning and orchestrated apps set: a run that produced
    stages and then stopped must not push the client's error branch and
    discard them. A refusal before stage 1 yields exactly one error event and
    nothing else, so a capped run never leaves a partial record.

    Args:
        payload: The validated request body.

    Returns:
        The event stream.
    """
    return EventSourceResponse(_run_stream(payload), ping=PING_SECONDS)


async def _run_stream(payload: RunRequest) -> AsyncGenerator[ServerSentEvent, None]:
    """Drive the run and put each stage on the wire as it completes.

    Args:
        payload: The validated request body.

    Yields:
        One SSE event per stage.
    """
    run_id = f"collab-{uuid.uuid4().hex[:16]}"
    delivered = 0

    async with async_session_factory() as session:
        stream = stream_run(
            session,
            run_id=run_id,
            scenario_id=payload.scenario_id,
            weighting_id=payload.weighting_id,
            weights=payload.weights,
        )
        try:
            async for event in stream:
                delivered += 1
                yield _event(event.kind, {"stage": event.stage, **event.payload})
        except asyncio.CancelledError:
            # Abandoning an `async for` does not close the inner generator, so
            # the sellers would keep bidding against a stream nobody reads --
            # spending quota on a run the visitor walked away from. Closing it
            # explicitly is what makes the stop immediate.
            await stream.aclose()
            logger.info("collab_run_abandoned", run_id=run_id, events=delivered)
            raise

    logger.info("collab_run_stream_closed", run_id=run_id, events=delivered)

File: frontend/src/api/collab.ts
// Built with Spec4 AI - https://spec4.ai
import { fetchEventSource } from '@microsoft/fetch-event-source'

/**
 * Typed client for the multi-agent collaboration endpoints.
 *
 * One client per example app, matching `orchestrated.ts` and `planning.ts`,
 * including their error messages: `fetch` rejects identically for "the backend
 * isn't running" and "the browser blocked the response for CORS", so the
 * unreachable message names both rather than saying something unactionable.
 *
 * The identity cards are fetched rather than duplicated here. They are what the
 * negotiation is validated against server-side, so a frontend copy would be a
 * second source of truth free to drift from the one the run actually uses.
 *
 * These interfaces mirror the backend's A2A-shaped models, camelCase included —
 * that spelling is the protocol's, not a TypeScript convention, and translating
 * it here would hide the one thing the shape is meant to show.
 */

/** The organisation standing behind an agent. */
export interface AgentProvider {
  organization: string
  url: string | null
}

/** One capability an agent advertises it can perform. */
export interface AgentSkill {
  id: string
  name: string
  description: string
  tags: string[]
  examples: string[]
}

/**
 * Transport-level features an agent supports.
 *
 * Every flag is `false` for every agent here, and that is accurate rather than
 * unfinished: there is no network transport to stream over or push across.
 */
export interface AgentCapabilities {
  streaming: boolean
  pushNotifications: boolean
  stateTransitionHistory: boolean
}

/** What a peer publishes about itself, as A2A shapes it. */
export interface AgentCard {
  name: string
  description: string
  version: string
  protocolVersion: string
  url: string | null
  provider: AgentProvider
  capabilities: AgentCapabilities
  skills: AgentSkill[]
  /** This showcase's own field, not A2A's. Always `'none'`: every agent here is knowledge-only. */
  toolAccess: string
  defaultInputModes: string[]
  defaultOutputModes: string[]
}

/**
 * One participant: its bus address and presentation, plus its published card.
 *
 * `id`, `role` and `color` sit beside the card rather than inside it because an
 * `AgentCard` describes an agent without addressing one — the same split the
 * backend keeps so its protocol module stays a faithful statement of A2A.
 */
export interface IdentityCard {
  id: string
  role: string
  color: string
  card: AgentCard
}

/** Response body of GET /api/collab/identity-cards. */
export interface IdentityCardsResponse {
  agents: IdentityCard[]
}

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000'

/** A failed collaboration request, carrying the backend's own explanation where there was one. */
export class CollabRequestError extends Error {
  readonly code: string

  constructor(message: string, code: string, cause?: unknown) {
    super(message, { cause })
    this.name = 'CollabRequestError'
    this.code = code
  }
}

/**
 * `fetch` rejects identically whether the backend is down or the browser
 * blocked the response for CORS, so the message names both rather than saying
 * something the reader cannot act on. Matches `orchestrated.ts`.
 */
function unreachable(url: string, cause: unknown): CollabRequestError {
  return new CollabRequestError(
    `Could not reach the backend at ${url}. Either it isn't running, or this page's origin (${window.location.origin}) doesn't match the backend's CORS_ORIGIN setting.`,
    'unreachable',
    cause,
  )
}

/**
 * Fetch the three peer identity cards.
 *
 * Static server-side configuration: no model call, no quota, nothing the
 * visitor supplied. It answers on a deployment with no provider keys at all,
 * which is what makes it safe to render on page load.
 *
 * @returns The buyer and both rival sellers, each with its A2A-shaped card.
 * @throws CollabRequestError if the backend is unreachable or returns non-2xx.
 */
export async function fetchIdentityCards(): Promise<IdentityCardsResponse> {
  const url = `${API_BASE_URL}/api/collab/identity-cards`

  let response: Response
  try {
    response = await fetch(url)
  } catch (cause) {
    throw unreachable(url, cause)
  }

  if (!response.ok) {
    throw new CollabRequestError(
      `The backend returned ${response.status} for ${url}.`,
      'request_failed',
    )
  }

  return (await response.json()) as IdentityCardsResponse
}

// ---------------------------------------------------------------------------
// The negotiation run
// ---------------------------------------------------------------------------

/** One seller's offer at one stage, as the backend emits it. */
export interface BidEvent {
  stage: string
  seller_id: string
  unit_price: number
  quantity: number
  delivery_days: number
  warranty_months: number
  notes: string
  concessions_made: string[]
  /** Present only on a bid re-issued by the differentiation check. */
  reissued?: boolean
}

/** The buyer's targeted push on one seller, on one term. */
export interface CounterOfferEvent {
  seller_id: string
  targeted_term: string
  ask: string
  justification: string
}

/** The buyer's own score for one seller on one priority. */
export interface PriorityScoreEvent {
  seller_id: string
  priority: string
  score: number
  comment: string
}

/** The buyer's decision, with the working it committed to first. */
export interface AwardEvent {
  winner_id: string
  per_priority_scoring: PriorityScoreEvent[]
  rationale: string
  priority_references: string[]
  runner_up_note: string
}

/** One row of the server's message log. */
export interface PeerMessageRow {
  sequence: number
  timestamp: string
  sender: string
  recipient: string
  stage: string
  work_item: Record<string, unknown>
}

/** What a run's cost is declared to be, before it starts. */
export interface DeclaredBudget {
  total: number
  negotiation: number
  explanation: number
}

/**
 * A stage event from `POST /api/collab/run`.
 *
 * Discriminated on `kind`, which is also the SSE event name. Anything the
 * client does not recognise is ignored rather than crashed on, so the backend
 * can add a stage without breaking a deployed frontend.
 */
export type RunEvent =
  | {
      kind: 'quotation_request'
      stage: string
      request: { text: string; goods: string; baseline_requirement: string }
      model_calls: number
      declared_budget: DeclaredBudget
      sellers: string[]
    }
  | ({ kind: 'bid'; stage: string } & BidEvent)
  | { kind: 'counter_offers'; stage: string; offers: CounterOfferEvent[]; repairs: string[] }
  | { kind: 'routing'; stage: string; delivered: string[]; model_calls: number }
  | { kind: 'degraded'; stage: string; seller_id: string; status: string }
  | {
      kind: 'award'
      stage: string
      award: AwardEvent
      reconciled: boolean
      reconciliation_note: string
      model_calls_used: number
      negotiation_stage_calls: number
    }
  | ({ kind: 'reveal'; stage: string } & RevealPayload)
  | ({ kind: 'sensitivity'; stage: string } & SensitivityPayload)
  | {
      kind: 'message_log'
      stage: string
      messages: PeerMessageRow[]
      seller_to_seller_count: number
    }
  | {
      kind: 'error'
      stage: string
      code: string
      message: string
      outcome?: string
      remaining?: number
      cap?: number
      resets_at?: string
    }

/** The event names the server sends. Anything else is ignored, not crashed on. */
const RUN_EVENT_NAMES = new Set([
  'quotation_request',
  'bid',
  'counter_offers',
  'routing',
  'degraded',
  'award',
  'reveal',
  'sensitivity',
  'message_log',
  'error',
])

/** Options for {@link startNegotiation}. */
export interface StartNegotiationOptions {
  scenarioId: string
  weightingId: string
  onEvent: (event: RunEvent) => void
  signal?: AbortSignal
}

/**
 * Start a negotiation and deliver each stage event as it arrives.
 *
 * `@microsoft/fetch-event-source` rather than the browser's `EventSource`,
 * and that is not a preference: `EventSource` is GET-only and a run starts
 * from a POST body. **Do not add a GET variant of the endpoint** to work
 * around it — that forks the API for a transport detail.
 *
 * Three of the library's defaults are wrong for an endpoint that spends quota,
 * and all three are overridden here for the same reasons the planning and
 * orchestrated clients override them:
 *
 * - `openWhenHidden: true`, because the default drops the connection on a
 *   backgrounded tab and *reopens* it — paying for the run twice, silently,
 *   and only for visitors who switch tabs.
 * - `onopen` throws the backend's own explanation, so a 422 on the request
 *   body does not read as a transport fault.
 * - `onerror` rethrows. The default retries forever whenever it *returns*.
 *
 * @param options - The scenario and weighting to run, plus the event sink.
 * @returns A promise resolving when the stream closes.
 * @throws CollabRequestError if the backend is unreachable or rejects the request.
 */
export async function startNegotiation({
  scenarioId,
  weightingId,
  onEvent,
  signal,
}: StartNegotiationOptions): Promise<void> {
  const url = `${API_BASE_URL}/api/collab/run`

  await fetchEventSource(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'text/event-stream' },
    body: JSON.stringify({ scenario_id: scenarioId, weighting_id: weightingId }),
    signal,
    openWhenHidden: true,

    async onopen(response) {
      if (response.ok) {
        return
      }
      const body: unknown = await response.json().catch(() => null)
      const detail =
        typeof body === 'object' && body !== null && 'detail' in body
          ? String((body as { detail: unknown }).detail)
          : `The backend returned ${response.status} for ${url}.`
      throw new CollabRequestError(detail, 'request_failed')
    },

    onmessage(message) {
      if (!RUN_EVENT_NAMES.has(message.event)) {
        return
      }
      onEvent({ kind: message.event, ...JSON.parse(message.data) } as RunEvent)
    },

    onerror(cause) {
      if (cause instanceof CollabRequestError) {
        throw cause
      }
      throw unreachable(url, cause)
    },
  })
}

/** One party's axis-by-axis reveal entry. */
export interface RevealAxis {
  axis: string
  stance: string
  opening_value: number
  final_value: number
  binding_constraint: string | null
  explanation: string
}

/** One party's unsealed block. */
export interface RevealParty {
  party_id: string
  headline: string
  axes: RevealAxis[]
}

/** The post-award unsealing payload. */
export interface RevealPayload {
  parties: RevealParty[]
  /** True when the narration came from the deterministic template. */
  fallback_generated: boolean
  violations: string[]
}

/** The computed half of the sensitivity projection, alongside the narration. */
export interface ComputedProjection {
  original_weights: Record<string, number>
  alternative_weights: Record<string, number>
  alternative_label: string
  original_winner: string
  alternative_winner: string
  outcome: string
  decisive_axes: string[]
}

/** The priority-sensitivity payload. */
export interface SensitivityPayload {
  likely_winner: string
  decisive_dimensions: string[]
  narration: string
  confidence: string
  caveat: string
  fallback_generated: boolean
  violations: string[]
  computed: ComputedProjection | Record<string, never>
}
